File: crates/cxx-qt-lib/src/types/qtime.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/include/qtime.h"

#include "assertion_utils.h"

#include <cstdint>
// QTime has one int member
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/time/qdatetime.h?h=v5.15.6-lts-lgpl#n242
//
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/time/qdatetime.h?h=v6.2.4#n217
assert_alignment_and_size(QTime, alignof(std::int32_t), sizeof(std::int32_t));

static_assert(std::is_trivially_copyable<QTime>::value,
              "QTime must be trivially copyable!");

namespace rust {
namespace cxxqtlib1 {

QTime
qtimeInitDefault()
{
  return QTime();
}

QTime
qtimeInit(int h, int m, int s, int ms)
{
  return QTime(h, m, s, ms);
}

}
}

File: tests/qt_types_standalone/rust/src/qtime.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QTime;

#[cxx::bridge]
mod qtime_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qtime.h");

        type QTime = cxx_qt_lib::QTime;
    }

    extern "Rust" {
        fn construct_qtime() -> QTime;
        fn read_qtime(p: &QTime) -> bool;
        fn clone_qtime(p: &QTime) -> QTime;
    }
}

fn construct_qtime() -> QTime {
    QTime::new(1, 2, 3, 4)
}

fn read_qtime(s: &QTime) -> bool {
    s.hour() == 1 && s.minute() == 2 && s.second() == 3 && s.msec() == 4
}

fn clone_qtime(s: &QTime) -> QTime {
    s.clone()
}
